modules/effects/arrow-effects-rx2-data/src/main/kotlin/arrow/effects/rx2/FlowableK.kt

```diff
@@ -190,41 +186,75 @@ data class FlowableK<A>(val flowable: Flowable<A>) : FlowableKOf<A>, FlowableKKi
-        val conn = FlowableKConnection()
-        // On disposing of the upstream stream this will be called by `setCancellable` so check if upstream is already disposed or not because
-        // on disposing the stream will already be in a terminated state at this point so calling onError, in a terminated state, will blow everything up.
-        conn.push(FlowableK { if (!emitter.isCancelled) emitter.onError(OnCancel.CancellationException) })
-        emitter.setCancellable { conn.cancel().value().subscribe() }
-
-        fa(conn) { either: Either<Throwable, A> ->
-          either.fold({
-            emitter.onError(it)
-          }, {
-            emitter.onNext(it)
+        fa { either: Either<Throwable, A> ->
+          either.fold({ e ->
+            emitter.tryOnError(e)
+          }, { a ->
+            emitter.onNext(a)
             emitter.onComplete()
           })
         }
       }, mode).k()
 
     fun <A> asyncF(fa: FlowableKProcF<A>, mode: BackpressureStrategy = BackpressureStrategy.BUFFER): FlowableK<A> =
       Flowable.create({ emitter: FlowableEmitter<A> ->
-        val conn = FlowableKConnection()
-        // On disposing of the upstream stream this will be called by `setCancellable` so check if upstream is already disposed or not because
-        // on disposing the stream will already be in a terminated state at this point so calling onError, in a terminated state, will blow everything up.
-        conn.push(FlowableK { if (!emitter.isCancelled) emitter.onError(OnCancel.CancellationException) })
+        val dispose = fa { either: Either<Throwable, A> ->
+          either.fold({ e ->
+            emitter.tryOnError(e)
+          }, { a ->
+            emitter.onNext(a)
+            emitter.onComplete()
+          })
+        }.fix().flowable.subscribe({}, emitter::onError)
+
+        emitter.setCancellable { dispose.dispose() }
+      }, mode).k()
+
+    fun <A> cancelable(fa: ((Either<Throwable, A>) -> Unit) -> CancelToken<ForFlowableK>, mode: BackpressureStrategy = BackpressureStrategy.BUFFER): FlowableK<A> =
+      Flowable.create<A>({ emitter ->
+        val token = fa { either: Either<Throwable, A> ->
+          either.fold({ e ->
+            emitter.tryOnError(e)
+          }, { a ->
+            emitter.onNext(a)
+            emitter.onComplete()
+          })
+        }
+        emitter.setCancellable { token.value().subscribe({}, { e -> emitter.tryOnError(e) }) }
+      }, mode).k()
 
-        val dispose = fa(conn) { either: Either<Throwable, A> ->
+    fun <A> cancelableF(fa: ((Either<Throwable, A>) -> Unit) -> FlowableKOf<CancelToken<ForFlowableK>>, mode: BackpressureStrategy = BackpressureStrategy.BUFFER): FlowableK<A> =
+      Flowable.create({ emitter: FlowableEmitter<A> ->
+        val cb = { either: Either<Throwable, A> ->
           either.fold({
-            emitter.onError(it)
-          }, {
-            emitter.onNext(it)
+            emitter.tryOnError(it).let { Unit }
```

Comment:
Is the `Unit` necessary?

Reply:
Yes, `tryOnError` checks if the emitter is disposed first and returns a boolean. `onNext` and `onComplete` already checks isDisposed first but return void instead.